webpack.config.js

```diff
@@ -1,17 +1,79 @@
 const path = require("path");
 const webpack = require('webpack');
+const HTMLWebpackPlugin = require('html-webpack-plugin');
+const CompressionPlugin = require("compression-webpack-plugin");
+const SWPrecacheWebpackPlugin = require('sw-precache-webpack-plugin');
 
-const config = {
-  entry: {
-    app: [
-      './src/index.js'
-    ]
-  },
 
+
+
+//enviroment variable
+const DEVELOPMENT = process.env.NODE_ENV === 'development';
+const PRODUCTION = process.env.NODE_ENV === 'production';
+const TEST = process.env.NODE_ENV === 'test'
+
+const entry = PRODUCTION
+      ?     {
+              src:
+              [
+                './index.js'
+              ]
+            }
+      :
+              [
+                './index.js'
+              ];
+
+const plugins = PRODUCTION
+      ?     [
+        new webpack.optimize.CommonsChunkPlugin({name:'vendor', filename:'vendor.[hash:12].min.js'}),
+        new webpack.optimize.UglifyJsPlugin({
+          compress: {
+            warnings: false,
+            screw_ie8: true,
+            dead_code: true,
+            unused: true
+          }
+        }),
+        new HTMLWebpackPlugin({
+          template:'index-template.html'
+        }),
+        new CompressionPlugin({
+          asset: "[path].gz[query]",
+          algorithm: "gzip",
+          test: /\.js$|\.css$|\.html$/,
+          threshold: 10240,
+          minRatio: 0.8
+        }),
+        new SWPrecacheWebpackPlugin(
+          {
+            cacheId: 'shuffle-box',
+            filename: 'shuffle.js',
+            forceDelete: true,
+            minify: true,
+            skipWaiting: true,
+            verbose: true
+          }
+        )
+      ]
+    : [
+        new webpack.NamedModulesPlugin()
+      ];
+
+//define plugins for both envs
+plugins.push(
+  new webpack.DefinePlugin({
+    'process.env': { NODE_ENV: JSON.stringify(process.env.NODE_ENV)}
```

Comment:
On some systems the process object will behave funny consider something  like  a heroku or aws box.